src/lib/homestore_backend/pg_blob_iterator.cpp

```diff
@@ -0,0 +1,129 @@
+    auto pg = get_pg_metadata();
+    pg_id_ = pg->pg_info_.id;
+    repl_dev_ = static_cast< HS_PG* >(pg)->repl_dev_;
+    if (upto_lsn != 0) {
+        // Iterate all shards and its blob which have lsn <= upto_lsn
+        for (auto& shard : pg->shards_) {
+            auto sequence_num = home_obj_.get_sequence_num_from_shard_id(shard->info.id);
+            if (shard->info.lsn <= upto_lsn) { max_shard_seq_num_ = std::max(max_shard_seq_num_, sequence_num); }
+        }
+    } else {
+        max_shard_seq_num_ = pg->shard_sequence_num_;
+    }
+}
+
+PG* HSHomeObject::PGBlobIterator::get_pg_metadata() {
+    std::scoped_lock lock_guard(home_obj_._pg_lock);
+    auto iter = home_obj_._pg_map.begin();
+    for (; iter != home_obj_._pg_map.end(); iter++) {
+        if (iter->second->pg_info_.replica_set_uuid == group_id_) { break; }
+    }
+
+    RELEASE_ASSERT(iter != home_obj_._pg_map.end(), "PG not found replica_set_uuid={}",
+                   boost::uuids::to_string(group_id_));
+    return iter->second.get();
+}
+
+void HSHomeObject::PGBlobIterator::create_pg_shard_snapshot_data(sisl::io_blob_safe& meta_blob) {
+    auto pg = get_pg_metadata();
+    auto& pg_info = pg->pg_info_;
+    auto& pg_shards = pg->shards_;
+
+    flatbuffers::FlatBufferBuilder builder;
+    std::vector< std::uint8_t > uuid(pg_info.replica_set_uuid.size());
+    std::copy(pg_info.replica_set_uuid.begin(), pg_info.replica_set_uuid.end(), uuid.begin());
+    auto pg_entry = CreatePGInfoEntry(builder, pg_info.id, 0 /* priority*/, builder.CreateVector(uuid));
+
+    std::vector< ::flatbuffers::Offset< ShardInfoEntry > > shard_entries;
+    for (auto& shard : pg_shards) {
+        auto& shard_info = shard->info;
+        // TODO add lsn.
+        shard_entries.push_back(CreateShardInfoEntry(
+            builder, static_cast< uint8_t >(shard_info.state), shard_info.placement_group, shard_info.id,
+            shard_info.total_capacity_bytes, shard_info.created_time, shard_info.last_modified_time));
+    }
+    builder.FinishSizePrefixed(CreateResyncPGShardInfo(builder, pg_entry, builder.CreateVector(shard_entries)));
+    meta_blob = sisl::io_blob_safe{builder.GetSize()};
+    std::memcpy(meta_blob.bytes(), builder.GetBufferPointer(), builder.GetSize());
+}
+
+int64_t HSHomeObject::PGBlobIterator::get_next_blobs(uint64_t max_num_blobs_in_batch, uint64_t max_batch_size_bytes,
+                                                     std::vector< BlobInfoData >& blob_data_vec, bool& end_of_shard) {
+    end_of_shard = false;
+    uint64_t total_bytes = 0, num_blobs = 0;
+    while (true) {
+        auto r = home_obj_.query_blobs_in_shard(pg_id_, cur_shard_seq_num_, cur_blob_id_ + 1, max_num_blobs_in_batch);
+        if (!r) { return -1; }
+        auto& index_results_vec = r.value();
+        for (auto& info : index_results_vec) {
+            if (info.pbas == HSHomeObject::tombstone_pbas) {
+                // Skip deleted blobs
```

Comment:
not sure can we do it like this?

suppose leader has {1,1} to {5,5}, and {5,7} to {10,10}({shard_id, blob_id}) , last lsn is 120,  {5,6} is deleted at lsn 100. the log at leader has been compacted to 110.
follower has {1,1} to {6.0} and the last lsn is 80 , then if baseline resync occurs, the follower will never konw that {5,6} has been deleted since it is not aware fo lsn 100.

so I think here we should set some special data for the tombstone_pbas in the blob_info_vec which will be send to follower , so that the follower can identify that this blob is deleted.

another question, if GC happens and tombstone is also removed, so how can leader let the follower know this info when baseline resync happens.

pls correct me if i missunderstand anything

Reply:
We havent put LSN into each blob index so ATM it is a full resync --- i.e all existing data can be discard.  So there is no issue for deleted blob and no necessary for transferring tombstone.

Extending the discussion further ,  assuming we have LSN in blob index,  we can let follower to set its current LSN and leader will only send the [follower_lsn ,  snapshot_lsn] to follower.  In this case,  as you said , we care about blob deletion.   The trivial approach is leader send out active blob list in <shard_id =S , batch =0 > , follower mark all blobs not in active blob list as deleted.

I think we are not yet have solid thinking regarding the "incremental snapshot" especially with a good amount of reserved log entries.  Though personally I am loving it.

Reply:
Yes this is a valid scenario. We can do two ways.
1. Either follower see's a gap in blob-sequence and assume its deletion.
2. More safe approach is use scrubber. In scrubber leader sends the valid list, its crc across followers.  Follower can use this to delete.

Reply:
I think before we have the `incremental snapshot` feature , i.e only transfer diff between 2 snapshots,    we would better erase everything on receiver side as anyway we start from scratch